ext/intl/formatter/formatter_format.c

```diff
@@ -53,30 +53,55 @@ PHP_FUNCTION( numfmt_format )
 			case IS_DOUBLE:
 				type = FORMAT_TYPE_DOUBLE;
 				break;
-			EMPTY_SWITCH_DEFAULT_CASE();
+			case IS_STRING:
+				type = FORMAT_TYPE_DECIMAL;
+				break;
+			default:
+				zend_argument_type_error(1, "must be of type int|float|string, %s given", zend_zval_type_name(number));
 		}
 	}
 
+	// Avoid losing precision on 32-bit platforms where PHP's "long" isn't
+	// as long as the FORMAT_TYPE_INT64 which is requested.
+#if SIZEOF_ZEND_LONG < 8
+	if (Z_TYPE_P(number) == IS_STRING && type == FORMAT_TYPE_INT64) {
+		type = FORMAT_TYPE_DECIMAL;
+	}
+#endif
+
 	switch(type) {
 		case FORMAT_TYPE_INT32:
+		{
+			bool failed = true;
+			int64_t value_64 = zval_try_get_long(number, &failed);
+			if (failed || value_64 < -2147483648 || value_64 > 2147483647) {
+				zend_argument_value_error(object ? 1 : 2, "must be numeric and fit in 32 bits");
+				RETURN_THROWS();
+			}
 			convert_to_long(number);
-			formatted_len = unum_format(FORMATTER_OBJECT(nfo), (int32_t)Z_LVAL_P(number),
+			formatted_len = unum_format(FORMATTER_OBJECT(nfo), (int32_t)value_64,
 				formatted, formatted_len, NULL, &INTL_DATA_ERROR_CODE(nfo));
 			if (INTL_DATA_ERROR_CODE(nfo) == U_BUFFER_OVERFLOW_ERROR) {
 				intl_error_reset(INTL_DATA_ERROR_P(nfo));
 				formatted = eumalloc(formatted_len);
-				formatted_len = unum_format(FORMATTER_OBJECT(nfo), (int32_t)Z_LVAL_P(number),
+				formatted_len = unum_format(FORMATTER_OBJECT(nfo), (int32_t)value_64,
 					formatted, formatted_len, NULL, &INTL_DATA_ERROR_CODE(nfo));
 				if (U_FAILURE( INTL_DATA_ERROR_CODE(nfo) ) ) {
 					efree(formatted);
 				}
 			}
 			INTL_METHOD_CHECK_STATUS( nfo, "Number formatting failed" );
+		}
 			break;
 
 		case FORMAT_TYPE_INT64:
 		{
-			int64_t value = (Z_TYPE_P(number) == IS_DOUBLE)?(int64_t)Z_DVAL_P(number):Z_LVAL_P(number);
+			bool failed = true;
+			int64_t value = zval_try_get_long(number, &failed);
+			if (failed) {
+				zend_argument_value_error(object ? 1 : 2, "must be numeric");
```

Comment:
```suggestion
				zend_argument_type_error(object ? 1 : 2, "must be of type int when argument #2 ($format) is NumberFormatter::TYPE_INT64");
```

Reply:
Ah yes, well the argument number can be a sprintf placeholder
``zend_argument_type_error(object ? 1 : 2, "must be of type int when argument #%d ($format) is NumberFormatter::TYPE_INT64", object ? 2 : 3);

Reply:
Got it, thanks!